File: USACO/previous_contest/Dec2012/bronzeC++/scramble/scramble.cpp
/*
 * scramble.cpp
 *
 *  Created on: Dec 22, 2018
 *      Author: Stanley
 *
 * Converting to Java
 * Testing differences between languages
 */

#include <fstream>
#include <algorithm>
#include <list>
#include <string>

using namespace std;

struct Name {
	string val;
	bool reversed;
	int idx;
};

bool comp_name (const Name& a, const Name& b) {
	return a.val < b.val;
}

int main() {
	ifstream fin("scramble.in");
	ofstream fout("scramble.out");

	int name_count;
	fin >> name_count;

	Name tmp_name;
	list<Name> names;
	for (int i = 0; i < name_count; i++) {
		tmp_name.idx = i;
		fin >> tmp_name.val;

		sort(tmp_name.val.begin(), tmp_name.val.end());
		tmp_name.reversed = false;
		names.push_back(tmp_name);

		reverse(tmp_name.val.begin(), tmp_name.val.end());
		tmp_name.reversed = true;
		names.push_back(tmp_name);
	}
	names.sort(comp_name);
	// Stable sort - sorted name will always be before reversed name

	int sorted_pos[50000], reversed_pos[50000];
	int sorted_count = 0, reversed_count = 0;
	for (list<Name>::iterator it = names.begin(); it != names.end(); it++) {
		if (it->reversed == false) {
			sorted_count++;
			sorted_pos[it->idx] = reversed_count + 1;
		}
		else {
			reversed_count++;
			reversed_pos[it->idx] = sorted_count;
			// Don't add 1 to account for the sorted counterpart
		}
	}

	for (int i=0; i<name_count; i++) {
		fout << sorted_pos[i] << " " << reversed_pos[i] << endl;
	}

	fout.close();

	return 0;
}

File: USACO/practice/Dec2012/bronzeJava/scramble/src/Scramble.java
import java.io.*;
import java.util.*;

public class Scramble {
	public static void main(String[] args) throws IOException {
		Scanner sc = new Scanner(new File("scramble.in"));
		int nameCount = sc.nextInt();
		sc.nextLine();

		String tmpStr;
		LinkedList<Name> names = new LinkedList<Name>();
		char[] tmpArray;
		for (int i = 0; i < nameCount; i++) {
			tmpStr = sc.nextLine();
			
			tmpArray = tmpStr.toCharArray();
			Arrays.sort(tmpArray);
			tmpStr = new String(tmpArray);
			names.add(new Name(tmpStr, i, false));
			
			tmpStr = new StringBuffer(tmpStr).reverse().toString();
			names.add(new Name(tmpStr, i, true));
		}
		sc.close();
		
		Collections.sort(names);

		PrintStream fout = new PrintStream("scramble.out");
		int[] sorted_pos = new int[50000];
		int[] reversed_pos = new int[50000];
		int sorted_count = 0, reversed_count = 0;
		ListIterator<Name> listIter = names.listIterator();
		Name cur;
		do {
			cur = listIter.next();
			if (cur.reversed == false) {
				sorted_count++;
				sorted_pos[cur.idx] = reversed_count + 1;
			}
			else {
				reversed_count++;
				reversed_pos[cur.idx] = sorted_count;
				// Don't add 1 to account for the sorted counterpart
			}
		}
		while (listIter.hasNext());

		for (int i=0; i<nameCount; i++) {
			fout.println(sorted_pos[i] + " " + reversed_pos[i]);
		}
		fout.close();
	}

	static class Name implements Comparable<Name> {
		String val;
		int idx;
		boolean reversed;
		
		Name (String val, int idx, boolean reversed) {
			this.val = val;
			this.idx = idx;
			this.reversed = reversed;
		}

		public int compareTo(Name n) {
			return this.val.compareTo(n.val);
		}

	}
}
